File: Observer/java/ObserverPattern.java
package Observer.java;

import java.util.HashSet;
import java.util.Set;

/**
 * Created by hanmomhanda on 2016-02-25.
 */
public class ObserverPattern {

    public static void main(String[] args) {

        Subject daddy = new Subject("오명운");

        Child yuna = new Child("오윤아");
        Child yoonsuk = new Child("오윤석");

        daddy.addChild(yuna);
        daddy.addChild(yoonsuk);

        daddy.saySomething("뒤에 가서 봐라~");

        daddy.removeChild(yuna);

        daddy.saySomething("어허~");
    }
}

interface Observer {
    void notified();
}

class Child implements Observer {

    private String name;

    public Child(String name) {
        this.name = name;
    }

    @Override
    public void notified() {
        System.out.println(name + "이(가) 뒤에가서 의자에 앉습니다.");
    }
}

class Subject {
    private String name;
    private Set<Child> childSet = new HashSet<>();

    public Subject(String name) {
        this.name = name;
    }

    public void addChild(Child child) {
        childSet.add(child);
    }

    public void removeChild(Child child) {
        childSet.remove(child);
    }

    public void saySomething(String saySomething) {
        System.out.println("+++++" + name + " : " + saySomething + "+++++");
        for (Child child : childSet) {
            child.notified();
        }
    }
}

File: Observer/ecma2015/ObserverPattern.js
'use strict';
class Subject {
    constructor(name) {
        this.name = name || 'default';
        this.observers = new Map();
    }
    
    addObserver(observer) {
        this.observers.set(observer.name, observer);
    }
    
    removeObserver(observer) {
        this.observers.delete(observer.name);
    }
    
    saySomething(something) {
        console.log(`+++++ ${this.name} : ${something} +++++`);
        this.notifyAll();
    }
    
    notifyAll() {
        // 아래 코드는 브라우저에서 정상동작하지만,
        // node 5.9.0에서는 동작하지 않음
        //for (let [k, v] of this.observers) {
        //    v.notified();
        //}

        for (let observer of this.observers) {
            observer[1].notified();
        }

    }
}

class Observer {
    constructor(name) {
        this.name = name || 'default';
    }
    
    notified() {
        throw new Error("This method should be overridden.");
    }
}

class Child extends Observer {
    constructor(name) {
        super(name);
    }
    
    notified() {
        console.log(`${this.name}이(가) 뒤에 가서 의자에 앉습니다.`);
    }
}

(() => {
    let daddy = new Subject('오명운');
    let yuna = new Child('오윤아');
    let yoonsuk = new Child('오윤석');

    daddy.addObserver(yuna);
    daddy.addObserver(yoonsuk);

    daddy.saySomething('뒤에 가서 봐라~');

    daddy.removeObserver(yuna);

    daddy.saySomething('어허~');
})();